### PR TITLE
Adds aws.Config.EndpointResolver that provides the ability to augment the service specific EndpointResolvers.

### PR DESCRIPTION
**Pull request by skmcgrail**
Adds aws.Config.EndpointResolver that provides the ability to augment the service specific endpoint resolvers.

### CODE REVIEW

File: codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/EndpointGenerator.java [hunk 1]
```diff
@@ -134,6 +136,51 @@ public void run() {
...
+            writer.openBlock("if w.awsResolver == nil {", "}", () -> writer.write("goto fallback"));
+
+            writer.write("endpoint, err = w.awsResolver.ResolveEndpoint(ServiceID, region)");
+            writer.openBlock("if err == nil {", "}", () -> writer.write("return endpoint, nil"));
+            writer.write("");
+
+            writer.addUseImports(SmithyGoDependency.ERRORS);
+            writer.openBlock("if nf := (&$T{}); !errors.As(err, &nf) {", "}", SymbolUtils.createValueSymbolBuilder("EndpointNotFound",
+                    AwsGoDependency.AWS_CORE).build(), () -> writer.write("return endpoint, err"));
+            writer.write("");
```
Comment (jasdel) [inline codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/EndpointGenerator.java:159]: line length here probably throw checkstyle validation error. and little hard to read.

File: codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/EndpointGenerator.java [hunk 2]
```diff
@@ -134,6 +136,51 @@ public void run() {
...
+                writer.addUseImports(SmithyGoDependency.FMT);
+                writer.write("return endpoint, fmt.Errorf(\"default endpoint resolver provided was nil\")");
+            });
+            writer.write("return w.resolver.ResolveEndpoint(region, options)");
+        });
+
+        // Generate exported helper for constructing a wrapper around the AWS EndpointResolver type that is compatible
+        // with the clients EndpointResolver interface.
+        writer.writeDocs(String.format("%s returns an EndpointResolver that first delegates endpoint resolution "
+                + "to the awsResolver. If awsResolver returns `aws.EndpointNotFound` error, the resolver will use the "
```
Comment (jasdel) [inline codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/EndpointGenerator.java:172]: Doesn't look like `EndpointNotFound` is defined. The type should probably also have the suffix, `EndpointNotFoundError` for easier discoverability/consistency with others in SDK..

**Review by jasdel — CHANGES_REQUESTED** [review]

File: codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/EndpointGenerator.java (hunk 2 above)
Comment (skmcgrail) [reply codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/EndpointGenerator.java:172]: It's defined in the `aws` Go package.

File: codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/EndpointGenerator.java (hunk 1 above)
Comment (skmcgrail) [reply codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/EndpointGenerator.java:159]: It's under the max, but will change for readability.

**Review by jasdel — APPROVED** [review]

**Pull request merged**